rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -0,0 +1,1424 @@
+        self,
+        data_signature: Dict[Text, List[FeatureSignature]],
+        label_data: RasaModelData,
+        inverted_tag_dict: Optional[Dict[int, Text]],
+        config: Dict[Text, Any],
+    ) -> None:
+        super().__init__(name="DIET", random_seed=config[RANDOM_SEED])
+
+        self.config = config
+
+        self.data_signature = data_signature
+        self._check_data()
+
+        self.predict_data_signature = {
+            k: vs for k, vs in data_signature.items() if "text" in k
+        }
+
+        label_batch = label_data.prepare_batch()
+        self.tf_label_data = self.batch_to_model_data_format(
+            label_batch, label_data.get_signature()
+        )
+        self._num_tags = len(inverted_tag_dict) if inverted_tag_dict is not None else 0
+
+        # tf objects
+        self._tf_layers = {}
+        self._prepare_layers()
+
+        # tf training
+        self._set_optimizer(tf.keras.optimizers.Adam(config[LEARNING_RATE]))
+        self._create_metrics()
+        self._update_metrics_to_log()
+
+        self.all_labels_embed = None  # needed for efficient prediction
+
+    def _check_data(self) -> None:
+        if "text_features" not in self.data_signature:
+            raise ValueError(
+                f"No text features specified. "
+                f"Cannot train '{self.__class__.__name__}' model."
+            )
+        if self.config[INTENT_CLASSIFICATION]:
+            if "label_features" not in self.data_signature:
+                raise ValueError(
+                    f"No label features specified. "
+                    f"Cannot train '{self.__class__.__name__}' model."
+                )
+            if (
+                self.config[SHARE_HIDDEN_LAYERS]
+                and self.data_signature["text_features"]
+                != self.data_signature["label_features"]
+            ):
+                raise ValueError(
+                    "If hidden layer weights are shared, data signatures "
+                    "for text_features and label_features must coincide."
+                )
+
+        if self.config[ENTITY_RECOGNITION] and "tag_ids" not in self.data_signature:
+            raise ValueError(
+                f"No tag ids present. "
+                f"Cannot train '{self.__class__.__name__}' model."
+            )
+
+    def _create_metrics(self) -> None:
+        # self.metrics preserve order
+        # output losses first
+        self.mask_loss = tf.keras.metrics.Mean(name="m_loss")
+        self.intent_loss = tf.keras.metrics.Mean(name="i_loss")
+        self.entity_loss = tf.keras.metrics.Mean(name="e_loss")
+        # output accuracies second
+        self.mask_acc = tf.keras.metrics.Mean(name="m_acc")
+        self.response_acc = tf.keras.metrics.Mean(name="i_acc")
+        self.entity_f1 = tf.keras.metrics.Mean(name="e_f1")
+
+    def _update_metrics_to_log(self) -> None:
+        if self.config[MASKED_LM]:
+            self.metrics_to_log += ["m_loss", "m_acc"]
```

Comment:
please use constants for all these magic strings

Reply:
they are used only here, not sure that constants will simplify things here

Reply:
they are also used in the response_selector :-)

Reply:
they're different in there